### ISSUE TITLE
Add device-local machine learning for custom filtering

### ISSUE THREAD

**Issue by Jeehut** [Ranchero-Software/NetNewsWire#3174]
## The Problem Statement
I tend to have a lot of news in my full "Today" view as well as when I select a specific RSS feed or folder and while I do want to keep the RSS sources, I'm not interested in _all_ the different topics in each of them, or at least not always. 

## Suggested Solution
A user can create custom "smart feeds" (we could also introduce a different name like "trained feeds") with their custom icon (a subset of SF Symbols could be used here), custom name and custom set of RSS feeds (from their existing list). For example consider the smart feed "Apple Device Reviews" with macrumors.com and 9to5mac.com as sources.

Then, users can mark any article from anywhere as "belongs to Apple Device Reviews" and "doesn't belong to Apple Device Reviews". At the beginning (without any trained data), the smart feed would show all news from the selected RSS feeds. But over time, the machine learning model would get better and filter out the unrelated articles.

Users can create as many smart feeds as they want, but probably it makes sense to create one by default so it's easy for users to explore this new feature. I would suggest a "Positive Mood" smart feed as the default which we could pre-train with some English articles. Other ideas are welcome.

## Alternatives Considered
I did find [this feature request](https://github.com/Ranchero-Software/NetNewsWire/issues/1864) asking for a regex or keyword based filtering, but I think that would be cumbersome to setup and not practical for many. But I can also see that we support both ways long term, I think both have their uses.

## Future Direction
For the first version I think, the machine learning model should be kept device-local to keep the features scope moderate. In a future version, we could think about changes to sync them across devices somehow.

In case there's no sophisticated API for this feature from Apple directly (I know you can pre-train and use that, but fully train on the running app, I'm not sure), we could opt for a simple solution with a higher error rate and replace that with an improved one once such an API is available.

## Implementation Details
Obviously, we should see if any of Apples official APIs (e.g. [CoreML](https://developer.apple.com/documentation/coreml)) already provides enough to get this working. If not, we could also opt for [TensorFlow Lite](https://www.tensorflow.org/lite) or simpler approaches such as a [Naive Bayes Classifier](https://en.wikipedia.org/wiki/Naive_Bayes_classifier).

**Comment by brentsimmons** [comment]
Years ago I considered using Apple’s Latent Semantic Mapping framework for classification: https://developer.apple.com/documentation/latentsemanticmapping

So I totally understand the idea and how it could be cool. However — one of NetNewsWire’s core principles is that there are no algorithms (in the sense of social media algorithms that try to order and filter what you see).

We will certainly add filters and smart feeds in the future, but those will be defined by the user rather than via machine learning or other algorithmic smarts.

I get that this answer is surely disappointing, and I’m sorry about that. I encourage you to fork the NNW code and create another RSS reader with this feature. Even just as a proof of concept. I think it would be interesting and a cool technical achievement, worth showing the world.

**Comment by Jeehut** [comment]
> social media algorithms that try to order and filter what you see

I did not suggest anything like that. 

> but those will be defined by the user

That's exactly what I'm suggesting here.

So, I'm sorry but I can't understand your argument. What you say only confirms that this feature would be useful and you want to add it. I have the feeling that you seem to have a one-sided view on machine learning. You seem to think that every machine learning algorithm is a "social media algorithm" and that every machine learning algorithm is out of control of the user.  

I have a different opinion here and I'm really happy there's such a thing as "spam filters" which I can train and which help me keep my email inbox clean. Does it work 100% of the time? No, it has false positives and false negatives from time t time. But do I prefer it over rule-based filters? Hell yeah, it's much easier to use and more efficient than those.

That's the kind of thing I suggested, spam filter like custom filters. But it's your project so I habe to respect your decision and will do as you suggested and create a fork. Maybe the result will convince you that we're not talking about any evil "social media algorithms" here. 😁
